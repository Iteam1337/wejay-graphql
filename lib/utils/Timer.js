function Timer(callback, delay) {
  var timerId,
    start,
    remaining = delay

  this.pause = function() {
    clearTimeout(timerId)
    remaining -= new Date() - start
  }

  this.resume = function() {
    start = new Date()
    clearTimeout(timerId)
    timerId = setTimeout(callback, remaining)
  }

  this.stop = function() {
    clearTimeout(timerId)
  }

  this.resume()
}

export default Timer
