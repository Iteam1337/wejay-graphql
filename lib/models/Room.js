import pubsub from '../services/pubsub'
import Timer from '../utils/Timer'

export default class Room {
  constructor(name, id) {
    this.name = name
    this.users = []
    this.queue = []
    this.id = id
    this.userSongs = {}
    this.currentTrack = null
    this.history = []
    this.timer = null
  }

  startTimer() {
    if (!this.currentTrack || !this.currentTrack.duration) {
      return false
    }

    pubsub.publish('onNextTrack', {
      roomName: this.name,
      onNextTrack: this.currentTrack
    })

    // keep track on the song length and automatically switch to next when the song ends
    this.timer = new Timer(() => {
      this.next()
    }, this.currentTrack.duration)
  }

  join(user) {
    this.users.push(user)
    this.userSongs[user.id] = []
  }

  queueTrack(track) {
    this.queue.push(track)

    this.userSongs[track.user].push(track)

    if (!this.currentTrack) {
      this.next()
    }
  }

  next() {
    if (this.timer) {
      this.timer.stop()
    }

    if (this.currentTrack) {
      const currentUser = this.users.find(
        user => user.id === this.currentTrack.user
      )

      currentUser.lastPlay = new Date()

      this.history.push(this.currentTrack)
      this.currentTrack = null
    }

    if (this.queue.length) {
      this.currentTrack = this.queue[0]
      this.currentTrack.started = new Date().toISOString()
      this.queue = this.queue.slice(1)

      this.startTimer()
    }
  }
}
