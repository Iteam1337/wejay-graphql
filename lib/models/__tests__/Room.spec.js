import MockDate from 'mockdate'

import Timer from '../../utils/Timer'
import Room from '../Room'

jest.mock('../../utils/Timer', () =>
  jest.fn(() => ({
    stop: jest.fn()
  }))
)

describe('Room', () => {
  let room

  beforeEach(() => {
    room = new Room('test', '1')

    Timer.mockClear()
  })

  afterEach(() => {
    MockDate.reset()
  })

  it('can create a Room', () => {
    expect(room).toMatchSnapshot()
  })

  describe('#join', () => {
    it('pushes a new user', () => {
      room.join({ email: 'cookie@monster.com' })

      expect(room.users).toMatchSnapshot()
    })

    it('adds the user to userSongs', () => {
      room.join({ email: 'cookie@monster.com', id: 'test' })

      expect(room.userSongs.test).toBeDefined()
    })
  })

  describe('#queueTrack', () => {
    it('queues a track', () => {
      room.currentTrack = { name: 'test' }
      room.queueTrack({ name: 'test' })

      expect(room.queue).toMatchSnapshot()
    })

    it('calls next if no current track', () => {
      MockDate.set('2017-09-20')

      room.queueTrack({ name: 'test' })

      expect(room.currentTrack).toEqual({
        name: 'test',
        started: '2017-09-20T00:00:00.000Z'
      })
    })
  })

  describe('#next', () => {
    it('removes any set timers', () => {
      room.timer = new Timer()

      room.next()

      expect(room.timer.stop).toHaveBeenCalled()
    })

    it('removes current track and sends it to history', () => {
      room.currentTrack = { name: 'test' }

      room.next()

      expect(room.history).toEqual([{ name: 'test' }])
      expect(room.currentTrack).toBeNull()
    })

    it('sets the first item in the list as current', () => {
      MockDate.set('2017-09-20')
      room.queue = [
        {
          duration: 100,
          name: 'test'
        }
      ]

      room.next()

      expect(room.currentTrack).toEqual({
        duration: 100,
        name: 'test',
        started: '2017-09-20T00:00:00.000Z'
      })
    })

    it('removes first track in queue', () => {
      room.queue = [
        {
          duration: 100,
          name: 'test'
        }
      ]

      room.next()

      expect(room.queue).toEqual([])
    })

    it('sets a Timer', () => {
      room.queue = [
        {
          duration: 100,
          name: 'test'
        }
      ]

      room.next()

      expect(Timer.mock.calls[0][1]).toEqual(100)
    })

    it('does not set a Timer if no current track duration', () => {
      room.queue = [
        {
          name: 'test'
        }
      ]

      room.next()

      expect(Timer).not.toHaveBeenCalled()
    })
  })
})
