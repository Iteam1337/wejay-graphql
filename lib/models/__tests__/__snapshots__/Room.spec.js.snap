// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Room #join pushes a new user 1`] = `
Array [
  Object {
    "email": "cookie@monster.com",
  },
]
`;

exports[`Room #queueTrack queues a track 1`] = `
Array [
  Object {
    "name": "test",
  },
]
`;

exports[`Room can create a Room 1`] = `
Room {
  "currentTrack": null,
  "history": Array [],
  "id": "1",
  "name": "test",
  "queue": Array [],
  "timer": null,
  "userSongs": Object {},
  "users": Array [],
}
`;
